### DISCUSSION TITLE
are custom (function) types possible?

### DISCUSSION THREAD

**Discussion by samuelstroschein** [sinclairzx81/typebox#529] [Q&A] [answered]
## Problem 

Sometimes custom types (type casts) are easier to work:

```ts
// 1. Need to define pattern to enforce camelCase but it's also a template literal string
const T = Type.String({
pattern: "^(?:[a-z0-9]+(?:[A-Z][a-z0-9]+)*)\\.lintRule[A-Z][a-zA-Z0-9]*$",
}) as unknown as TTemplateLiteral<[TLiteral<`${string}.lintRule${string}`>]>,

// 2. Typing functions is complex and validation not required
const T = Type.Object({
  type: Type.Literal("MessageLintRule"),
  defaultLevel: Type.String(),
  message: Type.Function()
})
```

## Question 

Are custom (function) types possible? 

```ts
// 1. Need to define pattern to enforce camelCase but it's also a template literal string
const T = Type.String<`${string}.lintRule${string}`>({
pattern: "^(?:[a-z0-9]+(?:[A-Z][a-z0-9]+)*)\\.lintRule[A-Z][a-zA-Z0-9]*$",
}) 

// 2. Typing functions is complex and validation not required
const T = Type.Object({
  type: Type.Literal("MessageLintRule"),
  defaultLevel: Type.String(),
  message: Type.Function<(args: { path: string } => string) >()
})
```

**Comment by sinclairzx81** [comment]
@samuelstroschein Hi,

TypeBox supports Function and Constructor types, but they are not very useful for validation (as JavaScript provides no specification or metadata to check a functions arguments or return types). Rather TypeBox includes them for use in interface definitions (think IDL) for describing callable interface methods (see [this project](https://github.com/sinclairzx81/sidewinder) for an example of their usage)

```typescript
const Add = Type.Function([Type.Number(), Type.Number()], Type.Number()])

type Add = Static<typeof Add> // type Add = (arg0: number, arg1: number) => number
```
With regards to their construction, it's not possible to construct Functions using TypeScript syntax alone (as TS will erase types on compilation), so the following won't work.

```typescript
const T = Type.Function<(a: number, b: number) => number>() // types are erased on compile
```
But you can use Type.Unsafe which will infer as any type you give it...however you will lose the metadata to describe the function (which may be ok depending on your usecase). Information on using Unsafe can be found [here](https://github.com/sinclairzx81/typebox#types-unsafe)
```typescript
const T = Type.Unsafe<(a: number, b: number) => number>({ ... })
```
Hope this helps!
S

**Reply by samuelstroschein** [reply]
I think `Type.Unsafe` is what I was looking for! Will report tomorrow. If so, I'll mark your reply as answer. 

Thanks for the quick response!

**Reply by samuelstroschein** [reply]
`Type.Unsafe` is exactly what I was looking for!

**Reply by samuelstroschein** [reply]
@sinclairzx81 False positive, `Type.Unsafe` is not what I am looking for because I am running into [ValueCheck: Unknown type 'Unsafe'](https://github.com/sinclairzx81/typebox/issues/528) as well. 

For our use case, we don't care about correctly validating a prop. We just want the type of the prop. Ideally, we can simply do the following:

```ts
const T = Type.Object({
  type: Type.Literal("MessageLintRule"),
  message: Type.Unsafe<() => string>()
})
```

**Reply by samuelstroschein** [reply]
@sinclairzx81 Using `Type.Unsafe` works for inferred types but fails on validation which is unexpected. The name `Unsafe` implies that "don't care about this prop". The following simple example doesn't work. 

https://stackblitz.com/edit/typescript-y4fsgz?file=index.ts

```ts
import { Type, Static } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';

const T = Type.Unsafe<string>({ type: 'number' });
type T = Static<typeof T>;

// expect true or false but error is thrown "Unknown type 'Unsafe'" 
console.log(Value.Check(T, 4));

```

### Proposal 

Let unsafe be unsafe and never throw?

**Reply by sinclairzx81** [reply]
@samuelstroschein Yes, you can register unsafe types with the TypeRegistry. You will need to give the Unsafe type a `[Kind]` that matches the registered type name.

```typescript
import { Type, Kind, TypeRegistry } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

// Register Here
TypeRegistry.Set('UnsafeFunction', (schema, value) => typeof value === 'function')

const UnsafeFunction = Type.Unsafe<() => string>({ [Kind]: 'UnsafeFunction' })

const T = Type.Object({
  type: Type.Literal("MessageLintRule"),
  message: UnsafeFunction
})

// true
//
const R1 = Value.Check(T, { type: 'MessageLintRule', message: () => 'hello' })

// false
//
const R2 = Value.Check(T, { type: 'MessageLintRule', message: null })
```
The TypeRegister function will receive the unsafe schema (unused in the example above), you can apply functions to the schema and receive them during validate (if you need to include specific logic with the schema)

Alternatively, you can also use the FormatRegistry if you're only checking for string values.

```typescript
import { FormatRegistry } from '@sinclair/typebox'

FormatRegistry.Set('MessageLintRule', (value) => true /* check string value somehow */)

const T = Type.String({ format: 'MessageLintRule' })

Value.Check(T, 'some lintable value')
```
Hope this helps
S

**Reply by samuelstroschein** [reply]
@sinclairzx81 The API is too difficult/boilerplaty. I suppose there are internal reasons for this. Thanks for the reply!

**Reply by sinclairzx81** [reply]
@samuelstroschein Hi. Yeah, things are written this way to decouple schematics from the logic that validates those schematics. It's mostly done to keep functions off the schemas (which would invalidate them as JSON Schema) as well as to support multiple schema specifications in future. There's a new specification under development for JSON Type Definition [here](https://github.com/sinclairzx81/typebox/tree/master/examples/typedef) that's built entirely off these registries (for reference).

The recommendation for custom formats (and types) is to register them globally once during application start up, and forget about them. Note, TypeBox is a bit more orientated towards integration than it is providing convenience API's for some of this stuff. The verbosity generally provides more options for integrators (with the goal of integration to make a lot of this boilerplate disappear, leaving only Types)

Anyway, good luck!
S

**Reply by samuelstroschein** [reply]
To clarify, the use case is **not to validate** but get TS type inference.

Hence, JSON schema, registering types, etc. is all too complicated to type functions with TS.

**Reply by sinclairzx81** [reply]
@samuelstroschein 

> To clarify, the use case is not to validate but get TS type inference.

If you only need type inference,  the Unsafe type is all you need. You would only register the Unsafe type if you intend to validate data with it.

```typescript
const T = Type.Unsafe<{ x: 1, y: 2 }>({}) // only register with [Kind] if you intend to validate.

type T = Static<typeof T> // type T = { x: 1, y: 2 }
```

**Reply by samuelstroschein** [reply]
@sinclairzx81 

Clarification:  The use case is not to validate function props/validate certain props and get TS type inference for all props.

### Problem 

Registering the unsafe type is a) unexpected (coming form zod) and b) unsuitable because the Unsafe type would leak to external validators which all need to register the unsafe type.

### Proposal 

Have a type that is ignored for validation to increase the usability of typebox without the need to register types.

- Either introduce a new primitive like `Type.Infer()` or extend  `Type.Unsafe()` to not require the registration of the unsafe type.

**Reply by sinclairzx81** [reply, marked as answer]
Would the following work?

```typescript
const T = Type.Unsafe<{x: 1, y: 2}>(Type.Any()) // a TAny type that infers as { x: 1, y: 2 }

Value.Check(T, 1)                               // true
Value.Check(T, false)                           // true
Value.Check(T, () => true)                      // true
```

**Reply by samuelstroschein** [reply]
Works! 

Adding it to the docs would likely be a good idea. Many thanks!

<img width="1430" alt="CleanShot 2023-08-14 at 12 11 32@2x" src="https://github.com/sinclairzx81/typebox/assets/35429197/ec4fd982-940a-4b8e-b1f0-d3c9f324552d">
